### ISSUE TITLE
TypeScript Plugin

### ISSUE THREAD

**Issue by dummdidumm** [sveltejs/language-tools#580]
The language-server has quite a few features by now. What's missing for the complete developer experience and intellisense though is crossing the boundaries of Svelte and JS/TS files. This can be achieved through a [TypeScript plugin](https://github.com/microsoft/TypeScript/wiki/Writing-a-Language-Service-Plugin). The following features would be cool to have:

- Rename variables in Svelte files, too, when renaming starts in a TS/JS file
- Go to Svelte definition/file from a TS/JS file
- Find references
- Better diagnostics (#550)

**Comment by dummdidumm** [comment]
Dumping some thoughts:
- A TS plugin needs to be synchronous, meaning all methods like `getDiagnostics` need to be synchronous. This is not possible to do right now since we use `source-map` which asynchronously generates a `SourceMapConsumer`. So either we switch the library or we do some kind of "use latest processed if available"-logic. What we definitely cannot do is reuse the language server connection and do `requestX, await response`.
- Because we cannot use the language server like the LSP protocol intends. Maybe it's possible to make the TypeScriptPlugin from the language server available for direct use while making sure it uses the same instance as the LS.

**Comment by pushkine** [comment]
Hey I worked on a proof of concept today 

https://user-images.githubusercontent.com/30108880/107896725-6e56d180-6f37-11eb-8ac9-fff759572208.mp4

https://github.com/pushkine/ts-plugin-svelte/

Beside obvious bugs where everything gets stale this doesn't yet work with Components, might be a conflict with `declare global module "*.svelte"` or some tweaking to do with `svelte2tsx`

Looks good to me though, what do you think ?

**Comment by dummdidumm** [comment]
This is a nice POC!

I though about this topic a little more and here is my roadmap on how we could get to a plugin:
1. Create performance-tests (#676)
2. Replace async `source-map` with sync `sourcemap-codec` + own litte mapping function similar to how I did it in [this ESLint plugin PR](https://github.com/sveltejs/eslint-plugin-svelte3/pull/83). Optionally reuse that logic in that PR to simplify+fix our mapping logic (could solve #666)
3. Test if the changes in 2 did not hurt the performance using the tests implemented in 1. If there are no big performance hits, continue.
4. Make all TypeScriptPlugin function synchronous - should be easily doable then because the only reason it was asynchronous was because of `source-map`
5. Refactor out the functionality of autocompletion/rename etc of the TypeScriptPlugin into a separate package (which we hopefully do not have to deploy, only pull it out for internal reuse)
6. Create a svelte-typescript-plugin. Both this plugin and the language-server use the internal package to provide TS/JS intellisense
7. Incorporate the plugin into the VS Code extension. Add an option to enable the plugin (maybe off by default for a short beta time).

Steps 1/2 can be done in parallel, the others are best done one after another. If someone wants to help in one of these steps, please ping here so we can coordinate. Step 5 is best done by me or @jasonlyu123 because we know the code base best and this is a critical part.

**Comment by jasonlyu123** [comment]
I also prefer option 3. Maybe we can export some functions from the plugin package or create another utility package to reuse some logic. For example, svelte-sys, source mapping utility, and source mapping workarounds.

**Comment by sastan** [comment]
**First: Thanks for all the great work you are doing! You are creating a really good DX.**

Now to the question:

I have created a [typescript plugin](https://github.com/tw-in-js/typescript-plugin) that powers [Twind IntelliSense for VS Code](https://github.com/tw-in-js/vscode-twind-intellisense).

Is it possible or would it be possible to have the completions/diagnostics from that typescript plugin within svelte files using the same implementation? 

I'm happy to contribute if there needs something to be implemented here. Or could give me some pointers/ideas on how that could be implemented in userland?

If this is the wrong issue I will move that into a separate issue.

Thanks for your time.

**Comment by dummdidumm** [comment]
Thanks for the kind words! This question is unrelated to this issue. This issue is about implementing a TS plugin for Svelte. What you want is to hook into the Svelte language server. Please open a separate issue for that.

**Comment by dummdidumm** [comment]
There exists a TypeScript plugin now which comes packaged with the VS Code extension and which you need to enable [through the settings](https://github.com/sveltejs/language-tools/tree/master/packages/svelte-vscode#svelteenable-ts-plugin). It also is [available standalone as a npm package](https://www.npmjs.com/package/typescript-svelte-plugin) if you need to use it outside of VS Code.
The one limitation is that you need to save changes of Svelte files to disk before they can be picked up inside TS/JS files.

Please test out the plugin and provide feedback. If you find bugs, please open separate issues for them.

**Comment by NickDarvey** [comment]
>It is still possible to compile the code with Rollup despite the Typescript errors, however this breaks the build step when using Sapper and Webpack with Typescript.

https://github.com/sveltejs/svelte/issues/5817#issue-773790759

@dummdidumm should using this plugin resolve the 'has no exported member' issue with exported members from Svelte modules when building with webpack? Or is it only intended to help IDEs?

**Comment by dummdidumm** [comment]
Only IDEs can use plugins, they are not loaded by tsc

**Comment by Jojoshua** [comment]
Just following up on https://github.com/sveltejs/language-tools/issues/1413 . I am using v105.13.0 of the extension and I do have **svelte.enable-ts-plugin : true** but the issue still persists on my work machine. Home machine is working fine. Anything I do/provide to help diagnose the issue?
